refactor pipeline validation unit tests v1beta1

pritidesai opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

This change is part of the larger refactoring effort to simplify unit tests
by splitting one huge validate test function into its own unit tests.
The end goal is to move the individual validation functions into their own
packages or have them available as attributes of the fields they are validating
instead of defining them locally under each APIs such as `v1beta1`.

Most changes done in test file, split one huge test function into small tests and added more test cases.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

pritidesai commented:
@bobcatfish I have split one huge unit test into individual validations, if these changes look good, I will apply similar changes to `v1alpha1` 😱😓

vdemeester commented:
@pritidesai https://github.com/tektoncd/pipeline/pull/2537 made v1beta1 builder to ease the port. Also, this is gonna conflict hard with https://github.com/tektoncd/pipeline/pull/2577 (as I am porting all the tests to v1beta1, starting with the builders to limit the size).
Also https://github.com/tektoncd/pipeline/pull/2577 will make sharing code between the two a bit useless (or less important), so I think we should wait for it to get merged and https://github.com/tektoncd/pipeline/issues/2410 to be fixed to do too much.

pritidesai commented:
@vdemeester as much I feel I am starting to understand the whole structuring of unit tests, I dont think I do 😭 

[v1beta1](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1beta1/pipeline_validation_test.go#L19) pipeline validation unit tests do not use `tb` whereas [v1alpha1](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1alpha1/pipeline_validation_test.go#L23) does. Are we deprecating `tb` or adopting it back or its completely out of context here? 🤔

vdemeester commented:
> @vdemeester as much I feel I am starting to understand the whole structuring of unit tests, I dont think I do sob
> 
> [v1beta1](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1beta1/pipeline_validation_test.go#L19) pipeline validation unit tests does not use `tb` whereas [v1alpha1](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1alpha1/pipeline_validation_test.go#L23) does. Are we deprecating `tb` or adopting it back or its completely out of context here? thinking

Long term goal is to deprecate them (mainly because the benefit they bring might not be worth), see https://github.com/tektoncd/pipeline/pull/2537#issuecomment-623783166. But this is long-term view, and a PR like #2577 would be ever bigger (and would take way more time) if I hadn't ported the builder to `v1beta1` :sweat_smile:

pritidesai commented:
thanks @vdemeester, I am refactoring `v1alpha1` pipeline validation and will drop `tb` from existing test cases in [pipeline_validation_test.go ](https://github.com/tektoncd/pipeline/blob/master/pkg/apis/pipeline/v1alpha1/pipeline_validation_test.go#L23) at least we can start with one 😜

bobcatfish commented on pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:
```diff
@@ -106,653 +62,948 @@ func TestPipeline_Validate(t *testing.T) {
-				Tasks: []v1beta1.PipelineTask{
-					{Name: "foo"},
-				},
+				Description: "this is a valid pipeline with all possible fields initialized",
```
nice!!

bobcatfish replied:
i wonder if it makes sense to have a similar failure case pipelinespec, e.g. one with a lot of different failures (i think they'll mostly all be surfaced? not sure if the logic just bails early in that case 🤔 )

pritidesai replied:
yup it does 👍

bobcatfish replied:
interesting!! totally outside the scope of what you're doing, but i feel like it'd be nice to be able to build up as many errors as possible and give them all at once

bobcatfish commented on pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:
```diff
@@ -106,653 +62,948 @@ func TestPipeline_Validate(t *testing.T) {
+	tests := []struct {
+		name            string
+		ps              *v1beta1.PipelineSpec
+		failureExpected bool
```
this is definitely outside the scope of your changes b/c the tests were already here but I really recommend having completely different tests for failure/non-failure cases in general - makes them much easier to read and maintain!

e.g. if you have a test where all the pipelines are invalid, and one where they are all valid, makes it much easier to understand whats covered and whats not

bobcatfish replied:
in fact it looks at a glance like all the cases here might be invalid so maybe the flag isnt needed?

pritidesai replied:
yup, will create separate functions for failure tests vs non-failure tests.

bobcatfish commented on pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:
```diff
@@ -106,653 +62,948 @@ func TestPipeline_Validate(t *testing.T) {
+		tasks:           []v1beta1.PipelineTask{{Name: "_foo", TaskRef: &v1beta1.TaskRef{Name: "foo-task"}}},
+		failureExpected: true,
+	}, {
+		name:            "pipeline task with invalid task name (camel case)",
```
nice - much more descriptive than "invalid task name 2"!

bobcatfish commented on pkg/apis/pipeline/v1beta1/pipeline_validation_test.go:
```diff
@@ -106,653 +62,948 @@ func TestPipeline_Validate(t *testing.T) {
-						}},
-					}},
+		name: "invalid condition only resource -" +
+			" pipeline task condition referring to a resource which is missing from resource declarations",
```
thanks for adding the details 🙏

bobcatfish commented:
![image](https://user-images.githubusercontent.com/432502/81615624-e93efa80-93af-11ea-8283-85f361bbf74f.png)

pritidesai commented:
thanks @bobcatfish, I will create separate functions for failure tests vs non-failure tests and create a separate PR with refactoring `v1alpha1` pipeline validation 🥰

pritidesai commented:
@bobcatfish @vdemeester thanks for the discussion this morning, I have updated the PR.

@bobcatfish I have split positive and negative cases into separate functions naming them `_Success` and `_Failure`, couldnt come up with better naming than this 😞 let me know if it looks ok.

bobcatfish commented:
> @bobcatfish @vdemeester thanks for the discussion this morning, I have updated the PR.

For posterity: we went back and forth a bunch on the best way to test these, in the end we all agreed (correct me if i'm wrong!) that it did make sense to have these focused tests, but that also indicated that in the long run these functions belong in a different package (we all agree that testing unexported functions isnt a good idea - it usually indicates some potential refactoring!) but in the short term we're testing the unexported functions here so that we get get the improved tests but without having to fix everything at once

thanks again @pritidesai !

/lgtm

pritidesai commented:
thanks @bobcatfish for details, yup thats accurate 👍
